File: database/scripts/validate-pro-tips-csv.py
#!/usr/bin/env python3
"""Validate MachineFit PRO tips CSV before import."""
from __future__ import annotations

import csv
import json
import os
import re
import sys
from collections import Counter, defaultdict
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
MIGRATION = ROOT / "database/migrations/133_seed_foundation_brands_standard_types.sql"

MAX_BYTES = 5000
MAX_LINES = 30
EXCLUDED_BRANDS = {"BODYWEIGHT", "FREE_WEIGHT"}
EXPECTED_BRANDS = 29
EXPECTED_MACHINES_PER_BRAND = 80
EXPECTED_ROWS = EXPECTED_BRANDS * EXPECTED_MACHINES_PER_BRAND

REQUIRED_HEADERS = ("brand_code", "machine_name", "exercise_tip", "exercise_tip_en")


def utf8_len(text: str) -> int:
    return len(text.encode("utf-8"))


def line_count(text: str) -> int:
    if not text.strip():
        return 0
    return len([ln for ln in re.split(r"\r?\n", text) if ln.strip()])


def load_brand_codes_from_migration() -> set[str]:
    text = MIGRATION.read_text(encoding="utf-8")
    codes = set(re.findall(r"'([A-Z0-9_]+)',\s*\n\s*'\{", text))
    # Also catch UPDATE/INSERT single-line brand codes
    codes.update(re.findall(r"WHERE code = '([A-Z0-9_]+)'", text))
    codes.update(re.findall(r"INSERT INTO brands \(code[^)]*\)\s*\nSELECT\s*\n\s*'([A-Z0-9_]+)'", text))
    return {c for c in codes if c not in EXCLUDED_BRANDS}


def load_standard_machine_names_from_migration() -> set[str]:
    text = MIGRATION.read_text(encoding="utf-8")
    names: set[str] = set()
    for m in re.finditer(r"'STD_[A-Z0-9_]+',\s*'\{\"ko\":\"([^\"]+)\"", text):
        names.add(m.group(1))
    return names


def try_load_db_catalog() -> dict[tuple[str, str], str] | None:
    """Return map (brand_code, std_name_ko) -> machine_code from live DB."""
    env_path = ROOT / "backend/.env"
    if not env_path.exists():
        return None
    db_url = None
    for line in env_path.read_text(encoding="utf-8").splitlines():
        if line.startswith("DATABASE_URL="):
            db_url = line.split("=", 1)[1].strip().strip('"').strip("'")
            break
    if not db_url:
        return None
    try:
        import psycopg2  # type: ignore
    except ImportError:
        return None
    try:
        conn = psycopg2.connect(db_url)
    except Exception:
        return None
    cur = conn.cursor()
    cur.execute(
        """
        SELECT b.code,
               COALESCE(st.name->>'ko', m.name->>'ko') AS std_ko,
               m.code AS machine_code,
               m.name->>'ko' AS machine_name_ko
        FROM machines m
        JOIN brands b ON b.id = m.brand_id
        LEFT JOIN standard_machine_types st ON st.id = m.standard_type_id
        WHERE b.code NOT IN ('BODYWEIGHT', 'FREE_WEIGHT')
          AND m.is_active = TRUE
        """
    )
    mapping: dict[tuple[str, str], str] = {}
    full_name_map: dict[tuple[str, str], str] = {}
    for brand_code, std_ko, machine_code, machine_name_ko in cur.fetchall():
        if std_ko:
            mapping[(brand_code, std_ko)] = machine_code
        if machine_name_ko:
            full_name_map[(brand_code, machine_name_ko)] = machine_code
            if std_ko and machine_name_ko.endswith(std_ko):
                mapping.setdefault((brand_code, std_ko), machine_code)
    conn.close()
    return {"std": mapping, "full": full_name_map}


def parse_csv(path: Path) -> tuple[list[str], list[dict[str, str]]]:
    with path.open("r", encoding="utf-8-sig", newline="") as f:
        reader = csv.DictReader(f)
        if reader.fieldnames is None:
            raise ValueError("Missing header row")
        headers = [h.strip() for h in reader.fieldnames if h]
        rows = []
        for raw in reader:
            row = {k.strip(): (v or "").strip() for k, v in raw.items() if k}
            rows.append(row)
        return headers, rows


def main() -> int:
    if len(sys.argv) < 2:
        print("Usage: validate-pro-tips-csv.py <csv-path>", file=sys.stderr)
        return 2
    csv_path = Path(sys.argv[1])
    if not csv_path.exists():
        print(f"File not found: {csv_path}", file=sys.stderr)
        return 2

    headers, rows = parse_csv(csv_path)
    report: dict = {
        "file": str(csv_path),
        "headers": headers,
        "rowCount": len(rows),
        "errors": [],
        "warnings": [],
        "stats": {},
    }

    missing_headers = [h for h in REQUIRED_HEADERS if h not in headers]
    if missing_headers:
        report["errors"].append(f"Missing columns: {', '.join(missing_headers)}")

    known_brands = load_brand_codes_from_migration()
    known_std_names = load_standard_machine_names_from_migration()
    db_maps = try_load_db_catalog()

    dup_counter: Counter[tuple[str, str]] = Counter()
    brand_counts: Counter[str] = Counter()
    unknown_brands: Counter[str] = Counter()
    unknown_machine_names: Counter[str] = Counter()
    over_bytes_ko: list[dict] = []
    over_bytes_en: list[dict] = []
    over_lines_ko: list[dict] = []
    over_lines_en: list[dict] = []
    empty_ko: list[int] = []
    empty_en: list[int] = []
    unmatched_db: list[dict] = []

    for idx, row in enumerate(rows, start=2):
        brand = row.get("brand_code", "").strip().upper()
        machine = row.get("machine_name", "").strip()
        tip_ko = row.get("exercise_tip", "")
        tip_en = row.get("exercise_tip_en", "")

        if not brand:
            report["errors"].append(f"Row {idx}: empty brand_code")
            continue
        if not machine:
            report["errors"].append(f"Row {idx}: empty machine_name")
            continue

        dup_counter[(brand, machine)] += 1
        brand_counts[brand] += 1

        if brand in EXCLUDED_BRANDS:
            report["errors"].append(f"Row {idx}: excluded brand {brand}")
        elif brand not in known_brands:
            unknown_brands[brand] += 1

        if machine not in known_std_names:
            unknown_machine_names[machine] += 1

        if not tip_ko.strip():
            empty_ko.append(idx)
        if not tip_en.strip():
            empty_en.append(idx)

        ko_bytes = utf8_len(tip_ko)
        en_bytes = utf8_len(tip_en) if tip_en.strip() else 0
        ko_lines = line_count(tip_ko)
        en_lines = line_count(tip_en)

        if ko_bytes > MAX_BYTES:
            over_bytes_ko.append({"row": idx, "brand": brand, "machine": machine, "bytes": ko_bytes})
        if tip_en.strip() and en_bytes > MAX_BYTES:
            over_bytes_en.append({"row": idx, "brand": brand, "machine": machine, "bytes": en_bytes})
        if ko_lines > MAX_LINES:
            over_lines_ko.append({"row": idx, "brand": brand, "machine": machine, "lines": ko_lines})
        if en_lines > MAX_LINES:
            over_lines_en.append({"row": idx, "brand": brand, "machine": machine, "lines": en_lines})

        if db_maps is not None:
            key = (brand, machine)
            std_map = db_maps["std"]
            full_map = db_maps["full"]
            if key not in std_map and key not in full_map:
                unmatched_db.append({"row": idx, "brand": brand, "machine": machine})

    duplicates = [(k, c) for k, c in dup_counter.items() if c > 1]
    if duplicates:
        report["errors"].append(f"Duplicate brand+machine pairs: {len(duplicates)}")
        report["stats"]["duplicateExamples"] = [
            {"brand": k[0], "machine": k[1], "count": c} for k, c in duplicates[:20]
        ]

    wrong_brand_counts = {
        b: c for b, c in brand_counts.items() if c != EXPECTED_MACHINES_PER_BRAND and b not in EXCLUDED_BRANDS
    }

    report["stats"].update(
        {
            "expectedRows": EXPECTED_ROWS,
            "brandCount": len(brand_counts),
            "brandsWithWrongRowCount": wrong_brand_counts,
            "unknownBrands": dict(unknown_brands.most_common(20)),
            "unknownMachineNameCount": len(unknown_machine_names),
            "unknownMachineNameExamples": list(unknown_machine_names.keys())[:20],
            "emptyKoRows": len(empty_ko),
            "emptyEnRows": len(empty_en),
            "overBytesKo": len(over_bytes_ko),
            "overBytesEn": len(over_bytes_en),
            "overLinesKo": len(over_lines_ko),
            "overLinesEn": len(over_lines_en),
            "maxKoBytes": max((utf8_len(r.get("exercise_tip", "")) for r in rows), default=0),
            "maxEnBytes": max((utf8_len(r.get("exercise_tip_en", "")) for r in rows), default=0),
            "avgKoBytes": round(sum(utf8_len(r.get("exercise_tip", "")) for r in rows) / max(len(rows), 1)),
            "avgEnBytes": round(sum(utf8_len(r.get("exercise_tip_en", "")) for r in rows) / max(len(rows), 1)),
            "dbMatchChecked": db_maps is not None,
            "dbUnmatchedRows": len(unmatched_db),
        }
    )

    if len(rows) != EXPECTED_ROWS:
        report["errors"].append(f"Row count {len(rows)} != expected {EXPECTED_ROWS}")

    if wrong_brand_counts:
        report["errors"].append(f"Brands not exactly {EXPECTED_MACHINES_PER_BRAND} rows: {len(wrong_brand_counts)}")

    if unknown_brands:
        report["errors"].append(f"Unknown brand codes: {len(unknown_brands)}")

    if empty_ko:
        report["errors"].append(f"Empty exercise_tip rows: {len(empty_ko)}")

    if over_bytes_ko:
        report["errors"].append(f"exercise_tip exceeds {MAX_BYTES} bytes: {len(over_bytes_ko)} rows")
        report["stats"]["overBytesKoExamples"] = over_bytes_ko[:10]

    if over_bytes_en:
        report["errors"].append(f"exercise_tip_en exceeds {MAX_BYTES} bytes: {len(over_bytes_en)} rows")
        report["stats"]["overBytesEnExamples"] = over_bytes_en[:10]

    if over_lines_ko:
        report["warnings"].append(f"exercise_tip exceeds {MAX_LINES} lines: {len(over_lines_ko)} rows")
        report["stats"]["overLinesKoExamples"] = over_lines_ko[:10]

    if over_lines_en:
        report["warnings"].append(f"exercise_tip_en exceeds {MAX_LINES} lines: {len(over_lines_en)} rows")
        report["stats"]["overLinesEnExamples"] = over_lines_en[:10]

    if empty_en:
        report["warnings"].append(f"Empty exercise_tip_en rows: {len(empty_en)} (will copy ko on import)")

    if unknown_machine_names:
        report["warnings"].append(
            f"machine_name not in standard 80-type catalog: {len(unknown_machine_names)} unique names"
        )

    if db_maps is not None and unmatched_db:
        report["errors"].append(f"Could not match to DB machines: {len(unmatched_db)} rows")
        report["stats"]["dbUnmatchedExamples"] = unmatched_db[:20]
    elif db_maps is None:
        report["warnings"].append("Live DB match skipped (no DATABASE_URL/psycopg2)")

    out_path = ROOT / ".cursor/handoff/pro-tips-csv-validation.json"
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(json.dumps(report, ensure_ascii=False, indent=2), encoding="utf-8")

    print(json.dumps(report, ensure_ascii=False, indent=2))
    return 0 if not report["errors"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: database/scripts/validate-pro-tips-csv.mjs
#!/usr/bin/env node
/**
 * Validate MachineFit PRO tips CSV before import.
 * Usage: node database/scripts/validate-pro-tips-csv.mjs <csv-path>
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, '../..');
const MIGRATION = path.join(ROOT, 'database/migrations/133_seed_foundation_brands_standard_types.sql');

const MAX_BYTES = 5000;
const MAX_LINES = 30;
const EXCLUDED_BRANDS = new Set(['BODYWEIGHT', 'FREE_WEIGHT']);
const EXPECTED_BRANDS = 29;
const EXPECTED_MACHINES_PER_BRAND = 80;
const EXPECTED_ROWS = EXPECTED_BRANDS * EXPECTED_MACHINES_PER_BRAND;
const REQUIRED_TIP_HEADERS = ['brand_code', 'exercise_tip', 'exercise_tip_en'];
const MACHINE_NAME_HEADERS = ['machine_name', 'machine_name_ko'];
const GENERIC_PHRASES = [
  '정확한 세부 모델 미확인',
  '브랜드명만 변경',
  'exercise_guidance_only',
];
const META_STATUSES = new Set([
  'VERIFIED',
  'PARTIALLY_VERIFIED',
  'BRAND_MODEL_NOT_FOUND',
  'exercise_guidance_only',
]);

function parseFlagValue(prefix) {
  const hit = process.argv.find((a) => a.startsWith(prefix));
  return hit ? hit.slice(prefix.length) : undefined;
}

function resolveMachineName(row) {
  return (row.machine_name_ko ?? row.machine_name ?? '').trim();
}

function assertCsvHeaders(headers, report) {
  if (!MACHINE_NAME_HEADERS.some((h) => headers.includes(h))) {
    report.errors.push('Missing column: machine_name or machine_name_ko');
  }
}

function utf8Len(text) {
  return Buffer.byteLength(text ?? '', 'utf8');
}

function lineCount(text) {
  if (!text?.trim()) return 0;
  return text.split(/\r?\n/).filter((ln) => ln.trim()).length;
}

function loadBrandCodes() {
  const text = fs.readFileSync(MIGRATION, 'utf8');
  const codes = new Set();
  for (const m of text.matchAll(/'([A-Z0-9_]+)',\s*\n\s*'\{/g)) codes.add(m[1]);
  for (const m of text.matchAll(/WHERE code = '([A-Z0-9_]+)'/g)) codes.add(m[1]);
  for (const m of text.matchAll(
    /INSERT INTO brands \(code[^)]*\)\s*\nSELECT\s*\n\s*'([A-Z0-9_]+)'/g
  )) {
    codes.add(m[1]);
  }
  for (const c of codes) {
    if (EXCLUDED_BRANDS.has(c)) codes.delete(c);
  }
  return codes;
}

function loadStandardMachineNames() {
  const text = fs.readFileSync(MIGRATION, 'utf8');
  const names = new Set();
  for (const m of text.matchAll(/'STD_[A-Z0-9_]+',\s*'\{\"ko\":\"([^\"]+)\"/g)) {
    names.add(m[1]);
  }
  return names;
}

/** RFC4180-ish CSV parser with multiline quoted fields. */
function parseCsv(content) {
  const rows = [];
  let row = [];
  let field = '';
  let inQuotes = false;
  for (let i = 0; i < content.length; i++) {
    const ch = content[i];
    const next = content[i + 1];
    if (inQuotes) {
      if (ch === '"' && next === '"') {
        field += '"';
        i++;
      } else if (ch === '"') {
        inQuotes = false;
      } else {
        field += ch;
      }
      continue;
    }
    if (ch === '"') {
      inQuotes = true;
      continue;
    }
    if (ch === ',') {
      row.push(field);
      field = '';
      continue;
    }
    if (ch === '\n') {
      row.push(field);
      field = '';
      if (row.some((c) => c.length > 0)) rows.push(row);
      row = [];
      continue;
    }
    if (ch === '\r') continue;
    field += ch;
  }
  row.push(field);
  if (row.some((c) => c.length > 0)) rows.push(row);
  if (!rows.length) return { headers: [], records: [] };
  const headers = rows[0].map((h) => h.trim());
  const records = rows.slice(1).map((cells) => {
    const obj = {};
    headers.forEach((h, idx) => {
      obj[h] = (cells[idx] ?? '').trim();
    });
    return obj;
  });
  return { headers, records };
}

async function tryLoadDbCatalog() {
  const envPath = path.join(ROOT, 'backend/.env');
  if (!fs.existsSync(envPath)) return null;
  const envText = fs.readFileSync(envPath, 'utf8');
  const match = envText.match(/^DATABASE_URL=(.+)$/m);
  if (!match) return null;
  const dbUrl = match[1].trim().replace(/^["']|["']$/g, '');
  let pg;
  try {
    pg = await import('pg');
  } catch {
    return null;
  }
  const client = new pg.default.Client({ connectionString: dbUrl, ssl: { rejectUnauthorized: false } });
  try {
    await client.connect();
  } catch {
    return null;
  }
  const res = await client.query(`
    SELECT b.code,
           COALESCE(st.name->>'ko', m.name->>'ko') AS std_ko,
           m.code AS machine_code,
           m.name->>'ko' AS machine_name_ko
    FROM machines m
    JOIN brands b ON b.id = m.brand_id
    LEFT JOIN standard_machine_types st ON st.id = m.standard_type_id
    WHERE b.code NOT IN ('BODYWEIGHT', 'FREE_WEIGHT')
      AND m.is_active = TRUE
  `);
  await client.end();
  const std = new Map();
  const full = new Map();
  for (const row of res.rows) {
    const brand = row.code;
    if (row.std_ko) std.set(`${brand}\0${row.std_ko}`, row.machine_code);
    if (row.machine_name_ko) full.set(`${brand}\0${row.machine_name_ko}`, row.machine_code);
  }
  return { std, full };
}

function main() {
  const csvPath = process.argv.find((a) => !a.startsWith('--') && a.endsWith('.csv')) ?? process.argv[2];
  const brandFilter = parseFlagValue('--brand=')?.trim().toUpperCase();
  const singleBrand = process.argv.includes('--single-brand') || Boolean(brandFilter);
  const expectedRows = singleBrand ? EXPECTED_MACHINES_PER_BRAND : EXPECTED_ROWS;

  if (!csvPath) {
    console.error('Usage: node validate-pro-tips-csv.mjs <csv-path> [--single-brand] [--brand=CODE]');
    process.exit(2);
  }
  if (!fs.existsSync(csvPath)) {
    console.error(`File not found: ${csvPath}`);
    process.exit(2);
  }

  const raw = fs.readFileSync(csvPath, 'utf8').replace(/^\uFEFF/, '');
  const { headers, records: allRecords } = parseCsv(raw);
  const records = brandFilter
    ? allRecords.filter((r) => (r.brand_code ?? '').trim().toUpperCase() === brandFilter)
    : allRecords;
  const report = {
    file: csvPath,
    headers,
    rowCount: records.length,
    singleBrand,
    brandFilter,
    expectedRows,
    errors: [],
    warnings: [],
    stats: {},
  };

  for (const h of REQUIRED_TIP_HEADERS) {
    if (!headers.includes(h)) report.errors.push(`Missing column: ${h}`);
  }
  assertCsvHeaders(headers, report);

  const knownBrands = loadBrandCodes();
  const knownStdNames = loadStandardMachineNames();

  const dup = new Map();
  const brandCounts = new Map();
  const unknownBrands = new Map();
  const unknownMachineNames = new Set();
  const overBytesKo = [];
  const overBytesEn = [];
  const overLinesKo = [];
  const overLinesEn = [];
  let emptyKo = 0;
  let emptyEn = 0;
  let maxKoBytes = 0;
  let maxEnBytes = 0;
  let sumKoBytes = 0;
  let sumEnBytes = 0;
  const metaStatusCounts = {};
  let metaInvalid = 0;
  let verifiedMissingSource = 0;
  let verifiedMissingModel = 0;
  let notFoundWithModel = 0;
  const genericPhraseHits = [];

  for (let i = 0; i < records.length; i++) {
    const rowNum = i + 2;
    const row = records[i];
    const brand = (row.brand_code ?? '').trim().toUpperCase();
    const machine = resolveMachineName(row);
    const tipKo = row.exercise_tip ?? '';
    const tipEn = row.exercise_tip_en ?? '';

    if (!brand) report.errors.push(`Row ${rowNum}: empty brand_code`);
    if (!machine) report.errors.push(`Row ${rowNum}: empty machine_name`);

    const key = `${brand}\0${machine}`;
    dup.set(key, (dup.get(key) ?? 0) + 1);
    brandCounts.set(brand, (brandCounts.get(brand) ?? 0) + 1);

    if (EXCLUDED_BRANDS.has(brand)) report.errors.push(`Row ${rowNum}: excluded brand ${brand}`);
    else if (brand && !knownBrands.has(brand)) unknownBrands.set(brand, (unknownBrands.get(brand) ?? 0) + 1);

    if (machine && !knownStdNames.has(machine)) unknownMachineNames.add(machine);

    if (!tipKo.trim()) emptyKo++;
    if (!tipEn.trim()) emptyEn++;

    const koBytes = utf8Len(tipKo);
    const enBytes = utf8Len(tipEn);
    const koLines = lineCount(tipKo);
    const enLines = lineCount(tipEn);
    maxKoBytes = Math.max(maxKoBytes, koBytes);
    maxEnBytes = Math.max(maxEnBytes, enBytes);
    sumKoBytes += koBytes;
    sumEnBytes += enBytes;

    if (koBytes > MAX_BYTES) overBytesKo.push({ row: rowNum, brand, machine, bytes: koBytes });
    if (tipEn.trim() && enBytes > MAX_BYTES) overBytesEn.push({ row: rowNum, brand, machine, bytes: enBytes });
    if (koLines > MAX_LINES) overLinesKo.push({ row: rowNum, brand, machine, lines: koLines });
    if (enLines > MAX_LINES) overLinesEn.push({ row: rowNum, brand, machine, lines: enLines });

    const status = (row.verification_status ?? '').trim();
    if (status) {
      if (!META_STATUSES.has(status)) metaInvalid++;
      metaStatusCounts[status] = (metaStatusCounts[status] ?? 0) + 1;
      if (status === 'VERIFIED' && !(row.source_url ?? '').trim()) verifiedMissingSource++;
      if (status === 'VERIFIED' && !(row.verified_model ?? '').trim()) verifiedMissingModel++;
      if (status === 'BRAND_MODEL_NOT_FOUND' && (row.verified_model ?? '').trim()) notFoundWithModel++;
    }

    for (const phrase of GENERIC_PHRASES) {
      if (tipKo.includes(phrase) || tipEn.includes(phrase)) {
        genericPhraseHits.push({ row: rowNum, brand, machine, phrase });
      }
    }
  }

  const duplicates = [...dup.entries()].filter(([, c]) => c > 1);
  if (duplicates.length) {
    report.errors.push(`Duplicate brand+machine pairs: ${duplicates.length}`);
    report.stats.duplicateExamples = duplicates.slice(0, 20).map(([k, c]) => {
      const [brand, machine] = k.split('\0');
      return { brand, machine, count: c };
    });
  }

  const wrongBrandCounts = Object.fromEntries(
    [...brandCounts.entries()].filter(
      ([b, c]) => !EXCLUDED_BRANDS.has(b) && c !== EXPECTED_MACHINES_PER_BRAND
    )
  );

  report.stats = {
    ...report.stats,
    expectedRows,
    brandCount: brandCounts.size,
    brandsWithWrongRowCount: wrongBrandCounts,
    unknownBrands: Object.fromEntries(unknownBrands),
    unknownMachineNameCount: unknownMachineNames.size,
    unknownMachineNameExamples: [...unknownMachineNames].slice(0, 20),
    emptyKoRows: emptyKo,
    emptyEnRows: emptyEn,
    overBytesKo: overBytesKo.length,
    overBytesEn: overBytesEn.length,
    overLinesKo: overLinesKo.length,
    overLinesEn: overLinesEn.length,
    maxKoBytes,
    maxEnBytes,
    avgKoBytes: records.length ? Math.round(sumKoBytes / records.length) : 0,
    avgEnBytes: records.length ? Math.round(sumEnBytes / records.length) : 0,
    overBytesKoExamples: overBytesKo.slice(0, 10),
    overBytesEnExamples: overBytesEn.slice(0, 10),
    overLinesKoExamples: overLinesKo.slice(0, 5),
    overLinesEnExamples: overLinesEn.slice(0, 5),
    metaStatusCounts,
    metaInvalid,
    verifiedMissingSource,
    verifiedMissingModel,
    notFoundWithModel,
    genericPhraseHits: genericPhraseHits.slice(0, 20),
    genericPhraseHitCount: genericPhraseHits.length,
  };

  if (records.length !== expectedRows) {
    report.errors.push(`Row count ${records.length} != expected ${expectedRows}`);
  }
  if (!singleBrand && Object.keys(wrongBrandCounts).length) {
    report.errors.push(
      `Brands not exactly ${EXPECTED_MACHINES_PER_BRAND} rows: ${Object.keys(wrongBrandCounts).length}`
    );
  }
  if (unknownBrands.size) report.errors.push(`Unknown brand codes: ${unknownBrands.size}`);
  if (emptyKo) report.errors.push(`Empty exercise_tip rows: ${emptyKo}`);
  if (overBytesKo.length) report.errors.push(`exercise_tip exceeds ${MAX_BYTES} bytes: ${overBytesKo.length} rows`);
  if (overBytesEn.length) report.errors.push(`exercise_tip_en exceeds ${MAX_BYTES} bytes: ${overBytesEn.length} rows`);
  if (overLinesKo.length) report.warnings.push(`exercise_tip exceeds ${MAX_LINES} lines: ${overLinesKo.length} rows`);
  if (overLinesEn.length) report.warnings.push(`exercise_tip_en exceeds ${MAX_LINES} lines: ${overLinesEn.length} rows`);
  if (metaInvalid) report.errors.push(`Invalid verification_status values: ${metaInvalid}`);
  if (verifiedMissingSource) {
    report.errors.push(`VERIFIED rows missing source_url: ${verifiedMissingSource}`);
  }
  if (verifiedMissingModel) {
    report.errors.push(`VERIFIED rows missing verified_model: ${verifiedMissingModel}`);
  }
  if (notFoundWithModel) {
    report.errors.push(`BRAND_MODEL_NOT_FOUND rows with verified_model set: ${notFoundWithModel}`);
  }
  if (genericPhraseHits.length) {
    report.errors.push(`Generic/template phrase detected in tips: ${genericPhraseHits.length} hits`);
  }
  if (emptyEn) report.warnings.push(`Empty exercise_tip_en rows: ${emptyEn} (will copy ko on import)`);
  if (unknownMachineNames.size) {
    report.warnings.push(
      `machine_name not in standard 80-type catalog: ${unknownMachineNames.size} unique names`
    );
  }

  return (async () => {
    const dbMaps = await tryLoadDbCatalog();
    if (!dbMaps) {
      report.warnings.push('Live DB match skipped (no DATABASE_URL/pg)');
    } else {
      const unmatched = [];
      for (let i = 0; i < records.length; i++) {
        const rowNum = i + 2;
        const brand = (records[i].brand_code ?? '').trim().toUpperCase();
        const machine = resolveMachineName(records[i]);
        const key = `${brand}\0${machine}`;
        if (!dbMaps.std.has(key) && !dbMaps.full.has(key)) {
          unmatched.push({ row: rowNum, brand, machine });
        }
      }
      report.stats.dbMatchChecked = true;
      report.stats.dbUnmatchedRows = unmatched.length;
      report.stats.dbUnmatchedExamples = unmatched.slice(0, 20);
      if (unmatched.length) report.errors.push(`Could not match to DB machines: ${unmatched.length} rows`);
    }

    const outPath = path.join(ROOT, '.cursor/handoff/pro-tips-csv-validation.json');
    fs.mkdirSync(path.dirname(outPath), { recursive: true });
    fs.writeFileSync(outPath, JSON.stringify(report, null, 2), 'utf8');
    console.log(JSON.stringify(report, null, 2));
    process.exit(report.errors.length ? 1 : 0);
  })();
}

main();
